src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -2702,77 +2706,210 @@ def delete_artifact(self, artifact_id: UUID) -> None:
 
     # -------------------- Artifact Versions --------------------
 
+    def _get_or_create_artifact_for_name(
+        self, name: str, has_custom_name: bool
+    ) -> ArtifactSchema:
+        """Get or create an artifact with a specific name.
+
+        Args:
+            name: The artifact name.
+            has_custom_name: Whether the artifact has a custom name.
+
+        Returns:
+            Schema of the artifact.
+        """
+        with Session(self.engine) as session:
+            artifact_query = select(ArtifactSchema).where(
+                ArtifactSchema.name == name
+            )
+            artifact = session.exec(artifact_query).first()
+
+            if artifact is None:
+                try:
+                    with session.begin_nested():
+                        artifact_request = ArtifactRequest(
+                            name=name, has_custom_name=has_custom_name
+                        )
+                        artifact = ArtifactSchema.from_request(
+                            artifact_request
+                        )
+                        session.add(artifact)
+                        session.commit()
+                    session.refresh(artifact)
+                except IntegrityError:
+                    # We failed to create the artifact due to the unique constraint
+                    # for artifact names -> The artifact was already created, we can
+                    # just fetch it from the DB now
+                    artifact = session.exec(artifact_query).one()
```

Comment:
This is to catch a race condition:
- We first check whether that artifact exists: If yes, return it.
- If two concurrent processes both do this check at the same time, the artifact might not exist for both of them. Now, they both try to create the artifact. Because this is now handled as a unique constraint on the DB, the creation will fail for one of the processes. This process can now fetch the artifact that was created by someone else.